### DISCUSSION TITLE
ESP32 hardware UARTs and firmware inconsistencies

### DISCUSSION THREAD

**Discussion by Allan196** [micropython/micropython#13648] [ESP32]
Platform: ESP32-WROVER (firmware ESP32_GENERIC-SPIRAM-20230426-v1.20.0.bin)

Using Thonny REPL I can open the three hardware UARTs: UART0, UART1 and UART2 as shown below (my input in bold):

>>>MicroPython v1.20.0 on 2023-04-26; ESP32 module with ESP32
>>>Type "help()" for more information.

>>> **from machine import UART**
>>> **UART(0)**
UART(0, baudrate=115201, bits=8, parity=None, stop=1, tx=-1, rx=-1, rts=-1, cts=-1, txbuf=256, rxbuf=256, timeout=0, timeout_char=0)

>>> **UART(1)**
UART(1, baudrate=115201, bits=8, parity=None, stop=1, tx=10, rx=9, rts=-1, cts=-1, txbuf=256, rxbuf=256, timeout=0, timeout_char=0)

>>> **UART(2)**
UART(2, baudrate=115201, bits=8, parity=None, stop=1, tx=17, rx=16, rts=-1, cts=-1, txbuf=256, rxbuf=256, timeout=0, timeout_char=0)
>>>

My current project uses UART(0) and UART(1).

I have since tried 2 later firmware versions:
ESP32_GENERIC-SPIRAM-20231005-v1.21.0.bin
and
ESP32_GENERIC-SPIRAM-20231227-v1.22.0.bin

Both of these firmware versions result in obscure error messages when I use Thonny and type UART(1) or UART(2).
UART(0) is okay as expected.

This is a quote from the MicroPython documentation:

> The ESP32 has three hardware UARTs: UART0, UART1 and UART2. They each have default GPIO assigned to them, however depending on your ESP32 variant and board, these pins may conflict with embedded flash, onboard PSRAM or peripherals.

So is it just luck that v1.20.0 allows me to use UART(1)?

**Comment by glenn20** [comment]
With MicroPython v1.22.1 (ie. latest release), these work fine for me. ie. `UART(0)`, `UART(1)` and `UART(2)` print identical lines to those above on an ESP32-WROVER although I am testing using `mpremote`, mot Thonny.

**Reply by Allan196** [reply]
So here is what happened with MicroPython v1.22.1 (using Thonny again). Also I have no peripherals attached to their pins.

MicroPython v1.22.1 on 2024-01-05; Generic ESP32 module with SPIRAM with ESP32
Type "help()" for more information.
>>> **from machine import UART**

>>> 
>>> **UART(0)**
UART(0, baudrate=115942, bits=8, parity=None, stop=1, tx=-1, rx=-1, rts=-1, cts=-1, txbuf=256, rxbuf=256, timeout=0, timeout_char=0)

>>> **UART(1)**
UART(1, baudrate=115201, bits=8, parity=None, stop=1, tx=10, rx=9, rts=-1, cts=-1, txbuf=256, rxbuf=256, timeout=0, timeout_char=0)

>>> **UART(2)**
UART(2, baudrate=115201, bits=8, parity=None, stop=1, tx=17, rx=16, rts=-1, cts=-1, txbuf=256, rxbuf=256, timeout=0, timeout_char=0)

>>> **UART(1).any()**
0
>>> PROBLEM IN THONNY'S BACK-END: Exception while handling 'get_globals' (thonny.plugins.micropython.mp_back.ManagementError: Management markers missing).
See Thonny's backend.log for more info.
You may need to press "Stop/Restart" or hard-reset your MicroPython device and try again.


Process ended with exit code 1.

**Comment by Tangerino** [comment]
Can you go further and call, at least, the any() function for all three
uarts? I got error there!!
So opening works fine in v1.22


*Merci beaucoup, Muito obrigado, Thanks*

*Carlos TangerinoMobile: +33 6 82 05 55 18*

**Reply by glenn20** [reply]
`UART(0).any()` returns
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 1] EPERM: ESP_FAIL
```
which presumably is because I am running the repl on UART(0).

`UART(1).any()` and `UART(2).any()` both return `0`. I have no peripherals attached to their pins.

**Comment by Allan196** [comment]
More testing feedback.
The following worked fine with v1.20.0 but failed with v1.22.1

MicroPython v1.22.1 on 2024-01-05; Generic ESP32 module with SPIRAM with ESP32
Type "help()" for more information.
>>> **from machine import UART**

>>> **uart1 = UART(1, baudrate=9600, bits=8, parity=None, stop=1, tx=32, rx=35, rts=-1, cts=-1, txbuf=256,rxbuf=256, timeout=0, timeout_char=0)** 

Traceback (most recent call last):
 
 File "<stdin>", line 1, in <module>
ValueError: invalid pin
>>> 

Regarding the invalid pin error above, I am able to configure Pin 32 as an output and Pin 35 as an input with no errors.
So is there a problem specific to UARTs?

**Reply by ricksorensen** [reply]
With respect to last UART connection using GPIO32, GPIO35  are there conflicts with PSRAM usage?  I believe on the S3 GPIO32 on up are used for PSRAM/SPIRAM  config.  Maybe check schematic for SPI RAM configuration?

Good luck.

**Reply by Allan196** [reply]
I tried a number of tx and rx pins and the error is always ValueError: invalid pin.
I couldn't find an exact schematic but similar ones don't show any pin conflicts in my choice of pins for UART(1)
